### ISSUE TITLE
Questions and comments

### ISSUE THREAD

**Issue by StephanSchuster** [tizianodeg/gruenbeck_softliQ_SC#2]
Hi. Thanks for this useful integration. I own a SC18 (like you?) and I am currently testing your stuff. So far it seems to work. But I have a couple of questions (out of curiosity) and comments.

This is what I see in HA:

![Screenshot from 2024-11-03 09-48-06](https://github.com/user-attachments/assets/7b60da95-504f-4ed5-bc95-951e65062847)
![Screenshot from 2024-11-03 09-47-55](https://github.com/user-attachments/assets/25b64207-7d85-4ad0-baef-bebb8151be13)

Questions & comments:
* Do the screenshots look like expected to you in terms of sensors and values for an SC18?
* Out of [Grünbeck's API doc](https://github.com/home-assistant/home-assistant/files/3998132/Webserver_Dokumentation.pdf), I wonder why you  chose exactly the sensors you did? I mean there are some that are not included in your integration but in the doc and vice versa. Did you test them all and then included those which were working and useful (for you)? Just want to understand the background.
* I think "[Durchschnittsverbrauch der letzten 3 Tage](https://github.com/tizianodeg/gruenbeck_softliQ_SC/blob/main/custom_components/gruenbeck_softliQ_SC/sensor.py#L169)" should be in m³ and not in m³/h
* "Salzreichweite" is apparently not available for SC18. But is there anything we can do with the "Letzter Fehler ..." fields to trigger a notifications once salts gets empty?
* Is the "Letzter Fehler Tage alt" valid on an SC18? Yesterday evening it was 46.480, now it is 46.494. Looks like it is hours, not days. But still this would mean 46494/24=1937 days which says about 5 years ago I missed to refill salt? Yeah, maybe. Not sure.
* When I first added your integration, "Gesamtdurchfluss" and "Salzverbrauch pro Jahr" were disabled (as expected when looking at you code). They apparently don't do anything on SC18, right? The strange thing, however, was that when I activated these two sensors via UI, the values of e.g. "Gesamtverbrauch" dropped back to 0 and started from scratch. This made me a bit suspicious. Is this a bug or expected? I don't want to loose statistic data when in future I e.g. update your integration or enable/disable its sensors.
* As of now, all of your sensors go to long term statistics. Is this really intended? I mean there a quite a few static values in there.

**Comment by StephanSchuster** [comment]
Ah, and one more thing: Is your intention that we should mainly use "Gesamtverbrauch" and maybe "Aktueller Durchfluss" and ignore the other diagnostic sensors? Then maybe add some utility meters in HA (daily, weekly, monthly, yearly) and work with these?

**Comment by StephanSchuster** [comment]
Last but not least: how are you visualizing the values in a dashboard? I don't want to add water to the default HA energy dashboard as the functionality is limited and because I don't want to mix solar energy stuff with water. I am more thinking of showing different usage graphs via Apex charts. Would be curious what and how you are visualizing it.

**Comment by hass-tiziano** [comment]
Hey thanks for the feedback,
Great to know this was somehow useful for someone else.

I'm using the default Energy Dashboard. I find it pretty useful and it does not get messy with the solar energy. It is also actual in the development Schoepe and gets improved.
So this was my main intent of this integration. Before I used rest calls, where I was not able to add it to the Energy Dashboard 
![image](https://github.com/user-attachments/assets/c1c1a442-cfb3-45df-88f1-7496897c980d)

So this is why I added the total consumption by collecting the actual consumption by code.

I would have loved to use some values from the sc18 but they are updated only once a day. So not really useful to understand if the high consumption is due to the dishwasher or the shower.
There are also some counters available from sc18 like yesterday etc. But using the energy monitor it is quite easy to retrieve all aggregation and query time.

I tried a lot different keys available and also some of the hidden once where a code is needed.
This is mainly what I found of interest, but as you have seen I did not fully checked all of them.

For the error I don't have an asset jet also waiting for my salt to be empty to check if there is a difference between an resolved error and errors that are still "valid" I suppose that the days/hours remain 0 as long as the error is still present.

Will have a detailed check later.

**Comment by hass-tiziano** [comment]
One thing I forgot, the total consumption is resettled to 0 each time the integration starts. This I what you have seen by activating the hidden sensors. I thought to use a initial value from sc18 on startup, but it gets messy as soon as the counter is changed due to the way sc18 increases the values only once a day resulting in a very high consumption at a specific time or negative values.
Found out that the "resetting to zero" is quite common and works pretty well with the energy dashboard. This way it is so easy and dump, no memory etc. Only if it does not work you will loose detailed consumption. I need to just floor the value it in the next release.

**Comment by StephanSchuster** [comment]
Understood. Thanks @tizianodeg for the insights.

I meanwhile found my setup. Basically I now also added water to the default energy dashboard. Then I use its cards on my own dashboard and hide the water via css where I don't want it, e.g. distribution. The last hour I visualize with Apex Charts to get the "real time feelling".

Please keep up your valuable work. Again, your stuff is super helpful. Looking forward to new releases.

Besides the minor things I mentioned earlier, your integration is also missing an icon:

![image](https://github.com/user-attachments/assets/f94f8b24-a803-420f-a027-2c8fc20cf8de)
![image](https://github.com/user-attachments/assets/65c48468-3b48-4bd6-8ce3-c82dc25bc51a)

**Comment by hass-tiziano** [comment]
Hey @StephanSchuster , fyi I totally forgot to monitor the salt level by really looking into the bin. 
Just realised that my dishwasher was getting chalky.... the bin was empty with no errors shown in my softliq, neither UI, app or on the display. 
So even I had some salt empty events triggered in the past between 10t to 9t hours ahead, this time no errors at all... 
I guess it was empty for about some months now and the water softener was still regenerating and "working" but outputting water with 20° hardness.

So after unmounting and cleaning up the bin I needed to run several regenerations, readjusting the reference point to get my softliq running again (no recommendation to anyone to do it by himself, saving 200€ for the inspection is not worth it.)

long story short I would not give much trust to the error to monitor the salt level...  Maybe with the SC23 it works better....

However I did some of the improvements you suggested, thanks for that again. 
Open Points would be:
- Icons:  I need to create a PR with icons to the home assistants brands repository... need to trigger this separately.
- Events for Errors: I don't have a real opinion on this and did not look into ha's possibilities on how to register such en entity in code. I will try to first generate a last error hours == 0 automation with an alert with the last error text as for now.

**Comment by Ye4ck** [comment]
Einfachheitshalber schreib ich jetzt auf Deutsch, 
"Letzter Fehler" kann nur mit dem Passwort 189 auf "0" gesetzt werden, konnte es leider noch nicht über http befehl testen.
"Letzter Fehler Stunden alt" wird in der Integration dadurch aber nicht auf 0 gesetzt. (im Mux wird er zurück gesetzt)

Edit: Falls es jemand testen kann/möchte. Es müsste mit folgendem edit funktionieren: `id=xxx&code=189&edit=D_M_3_3>0
`
Danke für die Integration :)

**Comment by hass-tiziano** [comment]
Ja, allerdings setzt man damit den gesamten Fehlerspeicher zurück. 
Was nicht meine Intention war, ich möchte schon noch sehen das es Fehler gab. 
Alle Fehlercodes zurückzusetzen ist m.M. eher eine Aufgabe, die ein Installateur übernehmen sollte. 

Mir ging es mehr darum anzuzeigen, dass der Salztank leer ist und ob es aktuell ist oder nicht. 
Neben dem letzten Fehler gibt es den "letzter Fehler Stunden alt", die Idee war:
- wenn der Wert auf 0 steht ist der Fehler aktuell. 
- wenn er > 0 ist, wurde Salz nachgefüllt. 

Gleiches könnte für andere Fehler helfen zu verstehen, ob man noch eingreifen muss oder ob sich der Fehler Erledigt hat. 

Allerdings konnte ich das noch nicht beobachten, auch weil er bei mir die "Salztank Leer" Meldung nicht gemeldet hat, obwohl mein Salztank leer war und 20° dH Wasser aus der Leitung kam...

**Comment by Ye4ck** [comment]
So gesehen macht das dann natürlich Sinn. 👍

Ja leider haben viele SC18 das Problem das der Salz fehler nicht ausgelöst wird,
meine bringt den Fehler leider auch nicht immer.

**Comment by tizianodeg** [comment]
btw. I added chlorine current sensor back in the 0.3 version. 
It only spikes during the regeneration process, I suppose that it will be less higher if the salt tank is empty. 
Could be a good indicator to check if the salt tank is empty or the tablets are stuck in the tank.

If you can have an eye on your values would be help full, as the longterm stats will create avg. on old values.
<img width="2890" height="802" alt="Image" src="https://github.com/user-attachments/assets/e017b66f-16a6-4c59-b062-abe6c11a440e" />

**Comment by GenerationKrise** [comment]
Wenn ich das Gerät als "Interaktion" hinzufügen will, kommt beim Eingeben vom Gerätenamen und dem Host (= IP Adresse) eine Fehlermeldung: "Unerwarteter Fehler".

**Comment by tizianodeg** [comment]
@GenerationKrise can you crosscheck if you have a SC18 model?
If you have a newer generation you need to use the cloud integration => https://github.com/p0l0/hagruenbeck_cloud.
see: https://github.com/tizianodeg/gruenbeck_softliQ_SC/issues/1

**Comment by GenerationKrise** [comment]
@tizianodeg I did not realize that there is a non-cloud-version existing. 
Anyway, with the integration you have linked it works! 
Thank you so much!

**Comment by Poaser** [comment]
when i have customized the sensor.py , muxclient, strings and translations and added a value(D_D_2), i still get the value as unknown in homeassistant - any ideas ?

<img width="651" height="181" alt="Image" src="https://github.com/user-attachments/assets/46881f4a-aa17-4f71-aba9-762d6a6286a4" />

**Comment by tizianodeg** [comment]
But you get values if you query the url manually?
btw. maybe D_A_1_6 is even more what you are looking for?
see https://github.com/tizianodeg/gruenbeck_softliQ_SC/blob/main/Webserver_Dokumentation.pdf

**Comment by Poaser** [comment]
thanks for the quick answer, before I had only used a simple post python script and had gotten values - honestly I could not do anything with the comments in the manual, what was meant by that :) but probably I am looking for exactly that :)

**Comment by tizianodeg** [comment]
Sorry, you are trying to get 
D_D_2 Double [°dH] Sollwert Weichwasserhärte

I guess this value is not changing, it should be the configured value that was defined when installing.

Instead 
D_A_1_6 
Anzeigewert = Istwert Weichwasserhärte bei
Anlagendurchfluss
>0,00 m³/h
Anzeigwert = Sollwert Weichwasserhärte bei
Anlagendurchfluss
= 0 m³/h

should be more interesting.. not sure why i did not take it into consideration... 
I can add it if you have trouble. Was currently working on the icon issue...

**Comment by Poaser** [comment]
ah got it :) 

no its totally fine just focus on the icon . I will probably test it in the afternoon

**
just received an ok without any values for D_D_2 and D_A_1_6 - maybe thats the reason why you dont implement this value 

<img width="1853" height="412" alt="Image" src="https://github.com/user-attachments/assets/5352ad6f-a064-40c0-ba2a-f1995ae93c8a" />

**Comment by alexaecho** [comment]
Thank you for the great custom integration. It works quite well with my SC18.

Would it be possible to add an icon to the integration that is shown in the HA frontend? 
E.g. this one:
https://www.gruenbeck.de/fileadmin/user_upload/01_Bildcontent/Bilder/Allgemeines_Bildmaterial/GruenbeckLogoKreis_web_2.svg

**Comment by tizianodeg** [comment]
> ah got it :)
> 
> no its totally fine just focus on the icon . I will probably test it in the afternoon
> 
> ** just received an ok without any values for D_D_2 and D_A_1_6 - maybe thats the reason why you dont implement this value
> 
> <img alt="Image" width="1853" height="412" src="https://private-user-images.githubusercontent.com/121462056/468582767-5352ad6f-a064-40c0-ba2a-f1995ae93c8a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTM5NzY1NzIsIm5iZiI6MTc1Mzk3NjI3MiwicGF0aCI6Ii8xMjE0NjIwNTYvNDY4NTgyNzY3LTUzNTJhZDZmLWEwNjQtNDBjMC1iYTJhLWYxOTk1YWU5M2M4YS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNzMxJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDczMVQxNTM3NTJaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0yY2I5MDViNzU1Y2MwMjYyMjhjOGNlNDA5YzhmZjUzMWU1ZGZiYWIwODc0M2FkMjViNTM2MGQ1ZWMzMzJkNjk4JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.ULI6YbWXehbqZdfNgPf1wqEfbUPmQPTp0ecYeurljdQ">

Yes also for me it returns just "-" does not look to be use full.
<img width="612" height="427" alt="Image" src="https://github.com/user-attachments/assets/ca3a6ed4-8796-4a94-aa71-28badc201a0b" />

To be honest I focused more on the values used by the frontend http://yourip you can look them up in the /var.js file.
The file doe not mention those values. Not sure who and how the pdf documentation was created.

**Comment by TSNr1** [comment]
Hi @tizianodeg,
first of all, thanks a lot for this great and usefull Integration.
My question is, if there is a easy to use solution, to get the waterconsumtion hourly (or more often) in the default Energydashboard?
Or @hass-tiziano ist it possible that you describe your way to solving it?

**Comment by hass-tiziano** [comment]
Hey,

The integration calculates a total consumption value based on the current flow that you can use in the Energy Dashboard:
![image](https://github.com/user-attachments/assets/b9b16894-dfb3-4651-873a-7244702a591e)

After a while the Energy Dashboard will show hour, daily or monthly consumption based on the zoom factor:
![image](https://github.com/user-attachments/assets/088a6d78-2232-4a0e-8d6c-33959a17d7f8)

![image](https://github.com/user-attachments/assets/e3318446-ecb1-4266-a11b-b50cd1c4b739)

You can also use Meters or Statistic helper tools to calculate other metrics based on total consumption.

Best, 
Tiziano

**Comment by TSNr1** [comment]
Thanks a lot, i try it, but i dont get a graphic like yours.
But i don't know why.

That's why i'm asking.

I had these Data. 
<img width="383" height="1048" alt="Image" src="https://github.com/user-attachments/assets/392f09b3-cde3-46eb-abf9-456f3c5b4ffe" />


And choose these entity

<img width="521" height="840" alt="Image" src="https://github.com/user-attachments/assets/b4843f31-daad-4f2c-82fc-db611c716e99" />

But there is no graphic like yours.

<img width="1492" height="525" alt="Image" src="https://github.com/user-attachments/assets/fcbcd619-8e06-45df-aa80-18c3df86cd7e" />

**Comment by TSNr1** [comment]
Oh i think i solved the problem.

My Sc18 doesn't count. I don't know why. I restarted it. And now it shows the "aktuellen Durchfluss"

**Comment by MichaelFink** [comment]
Hey there!
I also have an SC18 and came across this integration. Works like a charm. Great work 👍.
But I'm wondering if the value you report as "total consumption / Gesamtverbrauch" is correct. I'm pretty sure this is only the total consumption of soft water (so the 0°dh water which comes out of the device). This then gets mixed with the regular water to achieve an overall lower water hardness. So the actual water consumption is higher than what the SC18 reports directly. 
Would it make sense to allow configuring the integration so that the user can either enter the ratio the soft water gets mixed with regular water or the incoming water hardness and final outgoing water hardness? With that the real total consumption could be calculated. I think this is also how the SD18 actually calculates the total water consumption. I'd be happy to take a stab at this and send a PR if you think this makes sense.

Best,
Michael

**Comment by tizianodeg** [comment]
Hi @MichaelFink ,

for sure i would love to review your PR!

Currently i was aiming to show the same values as in the "Weichwasserverbrauch" view in the Grünbeck UI. 
<img width="1125" height="2436" alt="Image" src="https://github.com/user-attachments/assets/40786337-2499-4669-8875-cc8b0d57e55a" />
<img width="1125" height="2436" alt="Image" src="https://github.com/user-attachments/assets/f302fca4-5f59-423d-be5e-50906cbd918a" />

But as you mentioned the consumed looks to only show the 0°d soft water not the mixed one. In the Grünbeck UI picture you can see the total consumption to be twice as high as the summed daily consumptions.

But I'm not sure where you could find the configured factor to use to calculate the real intersection used. 
In the pdf it is  mentioned in parameters only available for MC32.

**Comment by frabu-t8p** [comment]
Hi, I tried to add my softliQ SE 24 using your integration, but upon entering the IP of the device (definitely the correct one...) I get a message "Unexpected Error". Any idea what might cause this!

Thanks

**Comment by Ye4ck** [comment]
This integration is only for the SC model, SE Models are cloud based. there is a different HA Integration for cloud based.

https://github.com/p0l0/hagruenbeck_cloud

> Hi, I tried to add my softliQ SE 24 using your integration, but upon entering the IP of the device (definitely the correct one...) I get a message "Unexpected Error". Any idea what might cause this!
> 
> Thanks

**Comment by frabu-t8p** [comment]
Bummer! Sorry, yes just noticed and installed the correct repository.  Sorry about that

**Comment by MichaelFink** [comment]
Sorry @tizianodeg for the late reply - I was traveling last week. See comments below.


> Hi [@MichaelFink](https://github.com/MichaelFink) ,
> 
> for sure i would love to review your PR!
> 
> Currently i was aiming to show the same values as in the "Weichwasserverbrauch" view in the Grünbeck UI. 
> But as you mentioned the consumed looks to only show the 0°d soft water not the mixed one. In the Grünbeck UI picture you can see the total consumption to be twice as high as the summed daily consumptions.

Are you sure the total consumption is twice as high? Did you fully scroll to the left to see all values? For me the total consumption is exactly the sum of all daily values (since I was travelling for the last three weeks I have just one day where I used water - so that's easy to verify 😉). See my screenshot below.

<img width="823" height="869" alt="Image" src="https://github.com/user-attachments/assets/7422a6ee-8995-4b33-aa55-23bace2b8489" />

> 
> But I'm not sure where you could find the configured factor to use to calculate the real intersection used. In the pdf it is mentioned in parameters only available for MC32.

Yeah, I think for the SC models you can't get the value from the device. So I would just allow the user to configure the mixing ratio in the config of the integration. We have two options here:
1. We either allow the user to enter the measured water hardness *after* the softener. Together with the incoming water hardness (before the softener), we could calculate the mixing ratio. E.g. if the incoming hardness is 10°dh and the outgoing hardness is 5°dh, the mixing ratio is 1:1. The precision of course depends on how accurately the two values were measured.
2. We give the user the option to just enter the mixing ratio directly. Users can get this value by comparing the values of their water meter and the soft water total consumption shown in the Grünbeck app. For example is your meter shows a consumption of 100 liters and you used 50 liters of soft water, you again have a mixing ratio of 1:1. 
This is probably the more precise approach. Precision is higher the larger the sample window is (e.g. you could also compare the values for an entire week  or month).

I'm a bit more inclined towards option 2 because of the (probably) higher precision. WDYT?
